### PR TITLE
GFDL candidate branch to main (2023-07-03)

### PR DESCRIPTION
**Pull request by marshallward**
This PR includes several new features to the model, a major rewrite of the IO layer, and preparatory work for the inclusion of non-Boussinesq mode.

Major features are the submesoscale parameterization of Bodner et al., new eigenmode-based computation of wave speeds for internal tides, and the mesoscale eddy parameterzation of Zanna-Bolton (2020).

The internal IO layer has been redesigned to exclusively use the FMS2 API when this framework (infra) has been enabled.  Although this has been tested and reviewed with partners to the best of our ability, results should be carefully checked, especially those depending on restart inputs.

Several internal operations related to Boussinesq mode have been redesigned and refactored as part of an ongoing effort to implement a full non-Boussinesq solver.  A "semi-Boussinesq" mode has been introduced to define the current state of the existing non-Boussinesq mode.

Note:
- Dimensional scaling factors have been removed from restart files.  Although files have changed, it should not impact runs (outside of certain debugging tests).

----

Features
--------

- NOAA-GFDL/MOM6#352: Adds the Bodner et al. 2023 version of MLE  (@abodner)
    - NOAA-GFDL/MOM6#385: *+Fix problems in mixedlayer_restrat_Bodner (@Hallberg-NOAA)
- NOAA-GFDL/MOM6#354: wave structure computation into wave_speeds (@raphaeldussin)
    - NOAA-GFDL/MOM6#383: +Refactor internal_tides interface (@Hallberg-NOAA)
- NOAA-GFDL/MOM6#356: Implementation of Zanna-Bolton-2020 equation discovery model of mesoscale momentum fluxes. Combined commits. (@Pperezhogin)

Framework
---------

- NOAA-GFDL/MOM6#343: Removal of FMS1 I/O from FMS2 I/O infra (@marshallward)
- NOAA-GFDL/MOM6#355: POSIX: generic wrappers for all setjmp.h symbols (@marshallward)
- NOAA-GFDL/MOM6#376: FMS2: Replace time_interp_external IDs with a MOM-defined type (@marshallward)
- NOAA-GFDL/MOM6#378: Use fileset rather than threading for decompositon (@marshallward)
- NOAA-GFDL/MOM6#386: FMS2: Safe inspection of unlimited dim name (@marshallward)
- NOAA-GFDL/MOM6#394: FMS2: New interface to set/nullify_domain (@marshallward)

Refactor
--------

- NOAA-GFDL/MOM6#300: +Make units argument mandatory for get_param_real (@Hallberg-NOAA)
- NOAA-GFDL/MOM6#331: *+Refactor, correct and augment the MOM6 EoS code (@Hallberg-NOAA)
- NOAA-GFDL/MOM6#335: (*)Use conversion factor for masscello diagnostic (@Hallberg-NOAA)
- NOAA-GFDL/MOM6#336: +Remove rescaling factors from restart files (@Hallberg-NOAA)
- NOAA-GFDL/MOM6#339: +Add and use US%Pa_to_RL2_T2 and US%Pa_to_RLZ_T2 (@Hallberg-NOAA)
- NOAA-GFDL/MOM6#342: Correct MLD_EN_VALS rescaling (@Hallberg-NOAA)
- NOAA-GFDL/MOM6#345: Add better error handling to read_var_sizes (@Hallberg-NOAA)
- NOAA-GFDL/MOM6#346: [WIP]Fix PGI runtime issue with class(*) (@nikizadehgfdl)
- NOAA-GFDL/MOM6#348: Checksum unrescaled non-Boussinesq thicknesses (@Hallberg-NOAA)
- NOAA-GFDL/MOM6#349: +(*)Better initialization of non-Boussinesq models (@Hallberg-NOAA)
- NOAA-GFDL/MOM6#368: +Add thickness_to_dz, calc_derived_thermo and avg_specific_vol (@Hallberg-NOAA)
- NOAA-GFDL/MOM6#365: +*Redefine GV%Angstrom_H in non-Boussinesq mode (@Hallberg-NOAA)
- NOAA-GFDL/MOM6#369: +Set_interp_answer_date and REGRIDDING_ANSWER_DATE (@Hallberg-NOAA)
- NOAA-GFDL/MOM6#373: *+Revise non-Boussinesq find_coupling_coef calcs (@Hallberg-NOAA)
- NOAA-GFDL/MOM6#384: +Add fluxes%tau_mag and forces%tau_mag (@Hallberg-NOAA)

Compilation
-----------

- NOAA-GFDL/MOM6#362: Autoconf: External FMS build configuration  (@marshallward)
- NOAA-GFDL/MOM6#366: Autoconf: Explicit MOM_memory.h configuration (@marshallward)
- NOAA-GFDL/MOM6#370: Autoconf: Better Unicode Python support in makedep (@marshallward)
- NOAA-GFDL/MOM6#371: Autoconf: Fix Python test and allow configuration (@marshallward)

Maintenance
-----------

- NOAA-GFDL/MOM6#353: Profiling: subparameter parser support (@marshallward)
- NOAA-GFDL/MOM6#361: Revised PR from main to dev/gfdl (@marshallward)
- NOAA-GFDL/MOM6#388: github workflows: update to use actions/checkout@v3 (@adcroft)

Contributors:
  - @Hallberg-NOAA: 16
  - @marshallward: 12
  - @adcroft: 1
  - @Pperezhogin: 1
  - @raphaeldussin: 1
  - @abodner: 1
  - @nikizadehgfdl: 1

### CODE REVIEW

**Comment by marshallward** [comment]
Regressions are expected due to the following:

* `masscello` diagnostic calculation order of operations.  See https://github.com/NOAA-GFDL/MOM6/pull/335
* Enabling of Bodner et al. parameterization in tc2.a: https://github.com/NOAA-GFDL/MOM6/pull/352

**Review by kshedstrom — APPROVED** [review]
I approve this PR.

**Review by abozec — APPROVED** [review]
COAPS approves.

**Comment by jiandewang** [comment]
now in EMC we have three systems to test.
(A) UFS fully coupled on global domain with all sorts of resolutions and combination of components
(B) HAFS regional ocean solo runs with OBC
(C) HAFS-MOM6 coupled runs which is new to MOM6 community. Here is a short description of what's C from Bin Li @binli2337:
_The HAFS-MOM6 system includes regional fv3atm, regional mom6, ww3, cmeps and cdeps components. For mom6, it is compiled using memory/dynamic symmetric. The coupling is through nuopc_cap._

A and B works as expected (no answer change) but B is having trouble. We suspect it is related to 
https://github.com/NOAA-GFDL/MOM6/pull/384
as the error is pointing to 
https://github.com/mom-ocean/MOM6/pull/1603/files#diff-3410befea8914fc74dc1848231ac30c7c31c64354183c45859bf0ef9adfc42dcR905

@marshallward and @Hallberg-NOAA I have a run log file on GAEA at
/lustre/f2/dev/ncep/Jiande.Wang/For-Marshall, see line 377-380

File: config_src/drivers/nuopc_cap/mom_surface_forcing_nuopc.F90 [hunk 1]
```diff
@@ -891,8 +899,10 @@ subroutine convert_IOB_to_forces(IOB, forces, index_bounds, Time, G, US, CS)
                  G%mask2dCv(i,J)*forces%tauy(i,J)**2) / (G%mask2dCv(i,J-1) + G%mask2dCv(i,J))
 
       if (CS%read_gust_2d) then
+        forces%tau_mag(i,j) = CS%gust(i,j) + sqrt(taux2 + tauy2)
         forces%ustar(i,j) = sqrt(CS%gust(i,j)*Irho0 + Irho0*sqrt(taux2 + tauy2))
       else
+        forces%tau_mag(i,j) = CS%gust(i,j) + sqrt(taux2 + tauy2)
```
Comment (jiandewang) [inline config_src/drivers/nuopc_cap/mom_surface_forcing_nuopc.F90:905]: should this be forces%tau_mag(i,j) = CS%gust(i,j)_const + ....... ?

Comment (Hallberg-NOAA) [reply config_src/drivers/nuopc_cap/mom_surface_forcing_nuopc.F90:905]: Yes, @jiandewang, you are indeed correct.  This line should be `forces%tau_mag(i,j) = CS%gust_const + sqrt(taux2 + tauy2)`.  Sorry about this oversight.

**Comment by alperaltuntas** [comment]
@marshallward This PR will require is to update to a newer FMS tag. And we need to coordinate that with the CESM FV3 folks. So, it may take a bit longer for us to review this PR. We are still using 2021.03.01. Any recommendation on which FMS tag to switch to?

**Comment by marshallward** [comment]
@alperaltuntas The newest 2023.x releases currently pass our regression test, and we intend to update our regression suite to use 2023.02 once the last few codebases have been updated.

The FMS1 infra should continue to work up to 2023.01.01, so may be a safer choice if you want to retain the ability to use either FMS1 or FMS2.

Most of the development was done with 2022.02, so that is another option to consider.

If you can handle a bit of disruption, then I would suggest one of the 2023.x releases (probably 2023.01 or 2023.01.01).  I don't expect this to be a completely smooth transition, but it's probably better to suss out these problems as soon as possible, since I don't think we're ever going back to the FMS1 infra.

Also: If we find any more FMS bugs, I will recommend that we send fixes immediately to main for review, rather than going through the usual dev/gfdl -> main cycle.

File: config_src/drivers/nuopc_cap/mom_surface_forcing_nuopc.F90 (hunk 1 above)
Comment (jiandewang) [reply config_src/drivers/nuopc_cap/mom_surface_forcing_nuopc.F90:905]: just got confirmation from Bin Li: HAFS-MOM6 works fine now with the correction on the typo.

**Comment by alperaltuntas** [comment]
> @alperaltuntas "This PR will require is to update to a newer FMS tag." We didn't think this to be the case. We use an older FMS tag in FMS1 mode. Can you clarify?

We recently switched to the FMS2 API, and we are getting the following bld error when this PR gets merged:

> Error: Symbol ‘flush_file’ referenced at (1) not found in module ‘fms2_io_mod’

Hence the reason to update to a newer FMS that includes the `flush_file` subroutine.

**Comment by marshallward** [comment]
> We recently switched to the FMS2 API, and we are getting the following bld error when this PR gets merged:
> 
> > Error: Symbol ‘flush_file’ referenced at (1) not found in module ‘fms2_io_mod’
> 
> Hence the reason to update to a newer FMS that includes the `flush_file` subroutine.

I believe that FMS2 had removed `flush_file`, but later re-introduced it:

https://github.com/NOAA-GFDL/FMS/commit/c057b753e7519b2c0c205ece7d9424420e0e92fa

There was a roughly 18-month period where our FMS2 infra was still using FMS1's `mpp_flush`, but this was eventually updated to the FMS2 flush.  This might explain why our FMS1 testing passes as far back as 2019.01.03 (and may even pass as far back as xanadu).

Many previous FMS2 releases may not work with our existing FMS2 infra, since it has arguably been in a state of development for the last few years.  I think the problems might even be worse than just a missing `flush` function.

I'm not sure exactly how far back we can go with the FMS2 infra, but I'm assuming it's at least 2022.02, since that is where I did most of our testing.  From what @alperaltuntas is saying, 2021.03.01 may be a breaking point, and the actual range may be very narrow.

TLDR: I think that FMS compatibility will depend on whether you are using the FMS1 or FMS2 infra.  If using the FMS1 infra, there is probably a wide of compatibility - butt this stops at 2023.02, when the FMS1 API is dropped.*  If using the FMS2 infra, then it's going to be spotty, and perhaps 2021 will not work.

(*: There is a way to restore the FMS1 API in the newest versions, but I think it's best we not focus on this.)

File: config_src/drivers/nuopc_cap/mom_surface_forcing_nuopc.F90 (hunk 1 above)
Comment (marshallward) [reply config_src/drivers/nuopc_cap/mom_surface_forcing_nuopc.F90]: The corrected expression has now been merged into the PR.

**Comment by marshallward** [comment]
The PR has been updated to include a fix to FMS I/O in horizontal regridding.  See https://github.com/NOAA-GFDL/MOM6/pull/399 for more info.

It's probably advisable to rerun your tests.

**Comment by marshallward** [comment]
Ping @jiandewang @gustavo-marques @sanAkel any update?

**Comment by gustavo-marques** [comment]
Alper is handling this PR on our side and he is on vacation until Aug 7th.

**Comment by sanAkel** [comment]
@marshallward Sorry I didn't see this PR until now.

Will look into in next few days, expect by Monday. Thanks!

**Comment by jiandewang** [comment]
@marshallward your updated PR has no negative impact on EMC side, I will wait for Alper's testing results before click approval in case there is any new updating required fron NCAR.

**Review by alperaltuntas — APPROVED** [review]
All NCAR tests passed (with FMS 2023.01.01).

**Comment by jiandewang** [comment]
since no any further code modification is needed, approve from EMC side

**Review by jiandewang — APPROVED** [review]

**Review by sanAkel — APPROVED** [review]
Answers preserved for GEOS. @marshallward thanks for patiently waiting!

**Comment by marshallward** [comment]
Thanks as always to everyone.
